lib/logstash/outputs/elasticsearch/http_client/pool.rb

```diff
@@ -0,0 +1,337 @@
+module LogStash; module Outputs; class ElasticSearch; class HttpClient;
+  class Pool
+    class BadResponseError < Error; end
+    class NoConnectionAvailableError < Error; end
+    class HostUnreachableError < Error;
+      attr_reader :original_error, :url
+
+      def initialize(original_error, url)
+        @original_error = original_error
+        @url = url
+      end
+
+      def message
+        "[#{original_error.class}] #{original_error.message}"
+      end
+    end
+
+    class BadResponseError < Error;
+      attr_reader :code, :uri, :body, :response
+
+      def initialize(response)
+        @code = response.code
+        @uri = response.request.uri
+        @body = response.body
+        @response = response
+      end
+
+      def message
+        "Received HTTP code #{@code} while retrieving #{@uri} with body: #{@body}"
+      end
+    end
+
+    attr_reader :logger
+
+    DEFAULT_OPTIONS = {
+      :healthcheck_path => '/'.freeze,
+      :scheme => 'http',
+      :resurrect_interval => 5,
+      :auth => nil, # Can be set to {:user => 'user', :password => 'pass'}
+      :sniffing => false,
+      :sniffer_delay => 10,
+    }.freeze
+
+    def initialize(logger, adapter, urls=[], options={})
+      @logger = logger
+      @adapter = adapter
+      @urls = urls
+
+      DEFAULT_OPTIONS.merge(options).tap do |merged|
+        @healthcheck_path = merged[:healthcheck_path]
+        @scheme = merged[:scheme]
+        @resurrect_interval = merged[:resurrect_interval]
+        @auth = merged[:auth]
+        @sniffing = merged[:sniffing]
+        @sniffer_delay = merged[:sniffer_delay]
+      end
+
+      # Override the scheme if one is explicitly set in urls
```

Comment:
I agree. out of the scope of this PR, but we should revisit the scheme logic that went on